# Constructing ghosts

The purpose of this notebook is to provide a general algorithm for constructing ghosts assuming the Stark conjectures and the twisted convolution identity. We focus (for now) on the case of rank 1 ghosts. We also leave necromancy for future work. 

This also serves as a test of the software package Zauner.gp. 

To load PARI/GP togeher with the Jupyter kernel that powers this notebook, use Conda and type

`conda install pari_jupyter -c conda-forge`

As of this writing, this should install PARI/GP version 2.13.3 into your current virtual environment together with the Jupyter kernel. 

What defines a ghost to begin with? 
Conjecturally, a ghost can be specified by:
- An integer $d > 3$. This is the dimension of the underlying complex vector space, $\mathbb{C}^d$. 
- A quadratic form $Q$ of discriminant $\Delta := f^2 \Delta_0$ where $\Delta \big\vert (d+1)(d-3)$ and $\Delta_0$ is a fundamental discriminant. 

_Remark:_ While the ghost formula depends explicitly on $Q$, different $Q$ in the same class give ghosts on the same extended Clifford orbit. 
Therefore, we can choose $Q$ to be a reduced form wlog. 
This choice is motivated by the fact that non-reduced forms require greater computational resources in general.

_Remark:_ To define a SIC, we would also (conjecturally) need to specify a sign-switching Galois automorphism $g(\Delta_0) = -\Delta_0$. 
This could be specified by an explicit choice of minimal polynomial for the field containing the ghost as well as an interval containing a root. 

Because we need quadratic forms of a given discriminant, it is germane to list all forms of a fixed discriminant, so let's start with that. 

In [2]:
1+1

2

In [1]:
read("../src/Zauner.gp");

Zauner.gp  v0.0.1


In [60]:
?qdata

qdata =
  (d,Q)->my(L,A,a,b,c,w);L=stabilizer(Q);a=stabilizersl2order(L,d);A=L^a;[a,b,c]=Vec(Q);w=quadgen(b^2-4*a*c,'w);beta=(-b+2*w-(b^2%4))/(2*a);[A,beta]



In [151]:
\p100

   realprecision = 115 significant digits (100 digits displayed)


In [152]:
d = 15
Q = Qfb(3, -18, 11)
[A,beta] = qdata(d,Q)

[[3106, -2145; 585, -404], 3 + 1/3*w]

In [153]:
allnu = vector(d^2-1, k, nu(A,d,radix(k,[d,d]),beta) );

  ***   at top-level: allnu=vector(d^2-1,k,nu(A,d,radix(k,[d,d]),bet
  ***                                      ^-------------------------
  ***   in function nu: .../sqrt(d+1);return(real(f*shin(A,d,p,beta)))
  ***                                               ^------------------
  ***   in function shin: ...[2])/d;S=prod(k=1,#zvals,sds(zvals[k],bvals[k]
  ***                                                 ^---------------------
  ***   in function sds: ...2-3*beta+1)/(24*beta));c=ds(z+n+1,beta,1);a*b*
  ***                                                ^---------------------
  ***   in function ds: ...2]=[-z,-w1,-w2];);return(dsShift(z,w1,w2))
  ***                                               ^-----------------
  ***   in function dsShift: ...Pi*(z-w2)/w1))),1,return(dsInt(z,w1,w2)))
  ***                                                    ^----------------
  ***   in function dsInt: ...(a,b,c,d,EPS);EPS=1/10;a=dsPowerSeries(w,b1,b2
  ***                                          

In [73]:
sqnu = vecsort(apply(sqr,allnu));

In [78]:
v = vector(#sqnu - 1, k, (k+1)*sign(floor(1E30*(sqnu[k+1]-sqnu[k]))) )

[0, 0, 4, 0, 0, 7, 8, 0, 0, 11, 0, 0, 14, 0, 0, 17, 0, 0, 20, 0, 0, 23, 0, 0, 26, 0, 0, 29, 0, 0, 32, 0, 0, 35, 0, 0, 38, 0, 0, 41, 0, 0, 44, 0, 0, 47, 0, 0, 50, 0, 0, 53, 0, 0, 56, 0, 0, 59, 0, 0, 62, 0, 0, 65, 0, 0, 68, 0, 0, 71, 0, 0, 74, 0, 0, 77, 0, 0, 80, 0, 0, 83, 0, 0, 86, 0, 0, 89, 0, 0, 92, 0, 0, 95, 0, 0, 98, 0, 0, 101, 0, 0, 104, 0, 0, 107, 0, 0, 110, 0, 0, 113, 0, 0, 116, 0, 0, 119, 0, 0, 122, 0, 0, 125, 0, 0, 128, 0, 0, 131, 0, 0, 134, 0, 0, 137, 0, 0, 140, 0, 0, 143, 0, 0, 146, 0, 0, 149, 0, 0, 152, 0, 0, 155, 0, 0, 158, 0, 0, 161, 0, 0, 164, 0, 0, 167, 0, 0, 170, 0, 0, 173, 0, 0, 176, 0, 0, 179, 0, 0, 182, 0, 0, 185, 0, 0, 188, 0, 0, 191, 0, 0, 194, 0, 0, 197, 0, 0, 200, 0, 0, 203, 0, 0, 206, 0, 0, 209, 0, 0, 212, 0, 0, 215, 0, 0, 218, 219, 0, 0, 222, 0, 0]

In [80]:
?listpop

listpop(~list,{n}): removes n-th element from list. If n is omitted or greater 
than the current list length, removes last element.



In [128]:
addhelp(union,"union(a,prec): the union of the list of numbers a, where equality is measured by precision prec.")
union(a,prec) = 
{
    my( b, n);
    a = vecsort(a);
    b = List(a);
    
    n = 1;
    for( k=2, #b,
        if( a[k] - b[n] < prec, 
            listpop(b,n+1),
            n = n+1
        );
    );
    Vec(b)
}

(a,prec)->my(b,n);a=vecsort(a);b=List(a);n=1;for(k=2,#b,if(a[k]-b[n]<prec,listpop(b,n+1),n=n+1););Vec(b)

In [133]:
u = union(apply(sqr,allnu),1e-30);
#u

76

In [142]:
upol = prod(k=1,#u, x^2 - (d+1)*u[k]);

In [144]:
ucoef = vector( poldegree(upol)+1, k, polcoef( upol, k-1, x) );

12

In [147]:
vector(#ucoef,k,lindep([-ucoef[k],1,sqrt(coredisc((d+1)*(d-3)))/2]) )

[[-196570803304774833267010930737784267074009168130301909198, 154635804235980603601974967093839867003289398948496721070, -202769229404830401395101852657335135384892192508956000495]~, [1, 0, 0]~, [3282719651793661920461193506821939494583070030821333232, 74407666504403595414224878196816233673666372039861220270, -1024164102587437070974294803030962995156421152302709838227]~, [1, 0, 0]~, [-165949025810321987037396436900501452417096400500717815, -6773101665281810713187084102553121934842776735146445612042, -8425296013806601365370895104885846262430885613699981680371]~, [1, 0, 0]~, [4680353449523196894681381786443106336674279287104422, -9058867035636175000193729104537875011044446604954118861917, -50161912817638151702881994223472505115810231716700560822046]~, [1, 0, 0]~, [91414508683239124592042081709342227062354710117699, -31134344884490943086263265824831655644812901729476870649600, 141966906251870642468683633487928377154034734978877730726184]~, [1, 0, 0]~, [-60271351646994994168575488807554984

In [55]:
1.0*quadgen(5)

1.6180339887498948482045868343656381177

In [44]:
2-sqrt(5)

-0.23606797749978969640917366873127623544

In [51]:
sicpol = x^6 + (3-2*quadgen(5))*(x^4+x^2)+1

x^6 + (3 - 2*w)*x^4 + (3 - 2*w)*x^2 + 1

In [59]:
polroots(sicpol)

  ***   at top-level: polroots(sicpol)
  ***                 ^----------------
  *** polroots: incorrect type in roots (t_QUAD).

In [59]:
??minpoly

minpoly(A,{v = 'x}):

   minimal  polynomial  of A with respect to the variable v.,  i.e.   the monic
polynomial  P  of  minimal  degree   (in  the variable v)  such that P(A)  = 0.

   The library syntax is GEN minpoly(GEN A,  long v = -1) where v is a variable
number.



In [59]:
apply(sqr,ssu)

[1.0000000000000000000000000000000000000 + 0.E-38*I, -0.23606797749978969640917366873127623544 - 0.97173654351329135636572775178906512439*I, -0.23606797749978969640917366873127623544 + 0.97173654351329135636572775178906512439*I]~

In [38]:
apply(abs,ssu)

[1.0000000000000000000000000000000000000, 1.0000000000000000000000000000000000000, 1.0000000000000000000000000000000000000, 1.0000000000000000000000000000000000000, 1.0000000000000000000000000000000000000, 1.0000000000000000000000000000000000000]~

In [4]:
d = 4;
[Delta0,f] = D0(d);
allf = divisors(f);

fq = allf[1]; \\ a divisor of f.
[qb,p,c] = ghostbasis(d,fq^2*Delta0);

Q = qb[1]^1;

L = stabilizer(Q);
a = stabilizersl2order(L,d);
A = L^a;

{
    my(a,b,c);
    [a,b,c] = Vec(Q);
    w = quadgen(b^2-4*a*c,'w); \\ quadratic generator for disc_Q
    beta = (-b+2*w-(b^2%4))/(2*a) \\ a positive root of Q
}


print("\nd = ", d, ", Δ = ", Delta0,"\nf = ", allf, "\nConductor fq = ", fq, "\nQ basis = ", [qb,p,c], "\nQ = ", Q, "\nA = psl2word(", psl2word(A), ")\n");



d = 4, Δ = 5
f = [1]
Conductor fq = 1
Q basis = [[Qfb(1, 1, -1, 0.E-38)], [1], 1]
Q = Qfb(1, 1, -1, 0.E-38)
A = psl2word([1, 3, 3, 2])



In [5]:
beta

-1 + w

In [3]:
default(realprecision,200)

derivnum(X=a,expr,{ind=1}): numerical derivation of expr with respect to X at X 
= a. The order of derivation is given by parameter 'ind', which can be a 
vector.

time = 1 ms

In [23]:
subst(deriv(pollegendre(4,'x)),x,0.9)

6.0075000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

In [23]:
?pollegendre

pollegendre(n,{a='x},{flag=0}): legendre polynomial of degree n evaluated at a. 
If flag is 1, return [P_{n-1}(a), P_n(a)].



In [14]:
vector(#t, k, pollegendre(4,t[k],1))

[5/2*x^3 - 3/2*x, 35/8*x^4 - 15/4*x^2 + 3/8]

In [110]:
[t, w] = gausslaguerre(550);

time = 21,570 ms

In [117]:
tt = apply(x->1/exp(4*x)+1/2,t);


time = 12 ms

In [118]:
w*tt

0.69999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999979415714

In [112]:
w*tt

0.40365263767680592565892150063072062391232461835689520527366123688646859138453813477024279992592897135833094598763240308165373749929313449195978930549681790976534609200381107936952362699299559055499901

In [65]:
default(realprecision,200)

In [119]:
ghostunittest(dmin,dmax,eps=1E-16) = 
{
    my(t = 0, QQ, G, d=dmin, projtest, a, b, c, t1);
    
    while( (d <= dmax) && (t < eps), 
        QQ = allQ(d);
        print("Checking ",#QQ," cases for d = ",d,".");
        for( j=1, #QQ, 
            Q = QQ[j];
            gettime();
            G = ghost(d,Q);
            projtest = G^2-G;
            [a,b,c] = Vec(Q);
            t = sqrt(norml2(projtest));
            print("    Q = ⟨",a,",",b,",",c,"⟩, ","‖G²-G‖₂ = ",t,", time = ",gettime()," ms");
            if( t > eps, error("Ghost idempotency is outside of error tolerance."));
        );
        d = d+1;
    );
}

In [119]:
ghostunittest(4,9)

Checking 1 cases for d = 4.
    Q = ⟨1,1,-1⟩, ‖G²-G‖₂ = 5.964161004751676619 E-35, time = 488 ms
Checking 1 cases for d = 5.
    Q = ⟨1,2,-2⟩, ‖G²-G‖₂ = 1.6122064555711913870 E-31, time = 801 ms
Checking 1 cases for d = 6.
    Q = ⟨1,3,-3⟩, ‖G²-G‖₂ = 3.363991524326751195 E-25, time = 1201 ms
Checking 2 cases for d = 7.
    Q = ⟨1,2,-1⟩, ‖G²-G‖₂ = 8.558910528337933317 E-26, time = 2824 ms
    Q = ⟨1,4,-4⟩, ‖G²-G‖₂ = 6.897715677598807139 E-24, time = 1580 ms
Checking 2 cases for d = 8.
    Q = ⟨1,1,-1⟩, ‖G²-G‖₂ = 1.1067344295825869175 E-23, time = 3263 ms
    Q = ⟨1,5,-5⟩, ‖G²-G‖₂ = 4.012851837476386911 E-21, time = 2088 ms
Checking 2 cases for d = 9.
    Q = ⟨1,6,-6⟩, ‖G²-G‖₂ = 1.2323313216367722517 E-20, time = 2701 ms
    Q = ⟨2,6,-3⟩, ‖G²-G‖₂ = 6.996692515522264308 E-22, time = 4732 ms
time = 19,684 ms

In [30]:
sicdatatable(d) = 
{
    my(D,s,f,ff,Delta,v);
    [D,s] = coredisc( (d+1)*(d-3), 1);
    f = divisors(s);
    ff = apply(sqr,f);
    Delta = ff*D;
    v = vector( #f, k, quadclassunit(Delta[k]) );
    for(k = 1, #v,
        if(v[k][1]==1, 
            v[k][2] = [1];
            v[k][3] = [prinredqfb(Delta[k])];
        );
    );
    v;
    
    h = vector(#f, k, towerh(d, Delta[k]));
    print(v,"\n",h)
}

(d)->my(D,s,f,ff,Delta,v);[D,s]=coredisc((d+1)*(d-3),1);f=divisors(s);ff=apply(sqr,f);Delta=ff*D;v=vector(#f,k,quadclassunit(Delta[k]));for(k=1,#v,if(v[k][1]==1,v[k][2]=[1];v[k][3]=[prinredqfb(Delta[k])];););v;h=vector(#f,k,towerh(d,Delta[k]));print(v,"\n",h)

In [30]:
sicdatatable(8)

[[1, [1], [Qfb(1, 1, -1, 0.E-38)], 0.48121182505960344749775891342436842314], [1, [1], [Qfb(1, 5, -5, 0.E-38)], 1.9248473002384137899910356536974736925]]
[2, 1]


In [120]:
sicclass(d) = 
{
    my(D,s,f,ff,Delta,v,t,u,h,a,b);
    [D,s] = coredisc( (d+1)*(d-3), 1);
    f = divisors(s);
    ff = apply(sqr,f);
    Delta = ff*D;
    v = vector( #f, k, quadclassunit(Delta[k]) );
    for(k = 1, #v,
        if(v[k][1]==1, 
            v[k][2] = [1];
            v[k][3] = [prinredqfb(Delta[k])];
        );
    );
    t = sum( k=1, #f, v[k][1] ); \\ total number in dimension d
    u = vector(#f, k, quadunit(Delta[k]));
    a = trace(u)/2;
    b = round(apply(sqrt,apply(sqr,u-a)/D));
    for(k = 1, #u,
        if(norm(u[k]) == -1, u[k] = u[k]^2);
    );
    h = round(acosh((d-1)/2)/log(u[1]));
    [d,D,h,t,f,v,u,a,b]
}


(d)->my(D,s,f,ff,Delta,v,t,u,h,a,b);[D,s]=coredisc((d+1)*(d-3),1);f=divisors(s);ff=apply(sqr,f);Delta=ff*D;v=vector(#f,k,quadclassunit(Delta[k]));for(k=1,#v,if(v[k][1]==1,v[k][2]=[1];v[k][3]=[prinredqfb(Delta[k])];););t=sum(k=1,#f,v[k][1]);u=vector(#f,k,quadunit(Delta[k]));a=trace(u)/2;b=round(apply(sqrt,apply(sqr,u-a)/D));for(k=1,#u,if(norm(u[k])==-1,u[k]=u[k]^2););h=round(acosh((d-1)/2)/log(u[1]));[d,D,h,t,f,v,u,a,b]

In [121]:
sicclass(9)

[9, 60, 1, 2, [1], [[2, [2], [Qfb(2, 6, -3, 0.E-38)], 2.0634370688955605467272811726201318715]], [4 + w], [4], [1]]

time = 1 ms

In [182]:
w = quadgen(4*13)

w

In [183]:
1.0*w

3.6055512754639892931192212674704959463

In [184]:
w = quadgen(13)

w

In [190]:
(w-0/2)^2

3 + w

In [191]:
minpoly(w)

x^2 - x - 3

In [191]:
??quadgen

quadgen(D,{v = 'w}):

   Creates the quadratic number omega =  (a+sqrt{D})/2 where a = 0 if D = 0 mod
4,   a  =  1  if  D  = 1 mod 4,  so that (1,omega) is an integral basis for the
quadratic  order  of discriminant D.   D must be an integer congruent to 0 or 1
modulo 4,  which is not a square.   If v is given, the variable name is used to
display g else 'w' is used.

   ? w = quadgen(5, 'w); w^2 - w - 1
   %1 = 0
   ? w = quadgen(0, 'w)
    ***   at top-level: w=quadgen(0)
    ***                   ^----------
    *** quadgen: domain error in quadpoly: issquare(disc) = 1

   The library syntax is GEN quadgen0(GEN D, long v = -1) where v is a variable
number.

/*-- (type RETURN to continue) --*/
   When  v does not matter,  the function GEN quadgen(GEN D) is also available.

time = 2 ms

In [172]:
??imag

imag(x):

   Imaginary part of x.   When x is a quadratic number, this is the coefficient
of omega in the "canonical" integral basis (1,omega).

   ? imag(3 + I)
   %1 = 1
   ? x = 3 + quadgen(-23);
   ? imag(x) \\ as a quadratic number
   %3 = 1
   ? imag(x * 1.) \\ as a complex number
   %4 = 2.3979157616563597707987190320813469600

   The library syntax is GEN gimag(GEN x).

time = 1 ms

In [192]:
d=5
u = quadunit(d)
[u, real(u), (trace(u)-d%4)/2, imag(u)]

[w, 0, 0, 1]

In [196]:
imag(u)

1

In [164]:
minpoly(u)

x^2 - 8*x - 1

In [168]:
u = quadunit(5);
if( norm(u) == -1, u = u^2)
x = trace(u)/2;
y = imag(u)

1

In [171]:
x

3/2

In [170]:
minpoly(u)

x^2 - 3*x + 1

time = 1 ms

In [123]:
d=39
quadclassunit((d+1)*(d-3))

[4, [2, 2], [Qfb(13, 32, -8, 0.E-38), Qfb(9, 36, -4, 0.E-38)], 3.6368929184641336469673979271214179876]

time = 1 ms

In [124]:
someQ = [Qfb(-5,11,5), Qfb(-5,9,7), Qfb(-7,5,7), Qfb(-7,9,5)]

[Qfb(-5, 11, 5, 0.E-38), Qfb(-5, 9, 7, 0.E-38), Qfb(-7, 5, 7, 0.E-38), Qfb(-7, 9, 5, 0.E-38)]

In [136]:
someL = apply(x->stabilizer(x)^3, someQ)

[[433, -1120; -1120, 2897], [657, -1568; -1120, 2673], [1105, -1568; -1568, 2225], [657, -1120; -1568, 2673]]

time = 1 ms

In [137]:
apply(psl2word, someL)

[[0, 3, 3, 2, 4, 3, 2, 4, 2, -2], [0, 2, 4, 2, 3, 4, 2, 3, 3, -2], [0, 2, 2, 4, 3, 2, 4, 3, 2, 3, -1], [0, 3, 2, 3, 4, 2, 3, 4, 2, 2, -1]]

time = 1 ms

The first step is to compute $L = L(Q)$, a generator of the stability group of $Q$, i.e. the group of matrices $M \in \mathrm{SL}_2(\mathbb{Z})$ such that $MQM^T = Q$. If we can write an arbitrary $M$ in this group as $M =\pm L^k$ for $k\in \mathbb{Z}$, then $L$ is a valid generator and any choice (there are four) gives us a choice of $L(Q)$. (The freedom in choice is $\pm L,\pm L^{-1}$.)

Some unit tests for the double sine function using the exact values computed in Shintani's original paper:
T. Shintani, "On a Kronecker limit formula for real quadratic fields", J. Fac. Sci. Univ.
Tokyo 24 (1977), 167–199.

Shin-ya Koyama and Nobushige Kurokawa, "Values of the double sine function", Journal of Number Theory, 123 1 (2007), 204-223.
http://www1.tmtv.ne.jp/~koyama/recentpapers/values.pdf

and from
Nobushige Kurokawa and Masato Wakayama, "Algebraicity and transcendency of basic special values of Shintani’s double sine functions", Proceedings of the Edinburgh Mathematical Society 49, (2006) 361–366.
